src/sentry/monitors/rate_limit.py

```diff
@@ -0,0 +1,73 @@
+from typing import Type
+
+from django.core.cache import cache
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from sentry.models.project import Project
+from sentry.monitors.models import MonitorEnvironment
+from sentry.tasks.relay import schedule_invalidate_project_config
+
+# Determines how many check-ins per-minute will be allowed per monitor. This is
+# used when computing the QuotaConfig for the DataCategory.MONITOR (check-ins)
+#
+#
+# Each monitor should __really__ only be checking-in once per minute, but we
+# will give some lee-way to allow for customers whos monitors check-in early.
+ALLOWED_CHECK_INS_PER_MONITOR_PER_MINUTE = 5
+
+# The minimum computed project_monitor_rate_limit. This value should be high
+# enough that it allows for a large number of monitors to be upserted without
+# hitting the project rate-limit.
+ALLOWED_MINIMUM = 50
+
+
+def get_project_monitor_rate_limit(project: Project, cache_bust=False) -> int:
+    """
+    Determines the rate-limit for monitor check-ins across a particular
+    project.
+
+    This value is intended to be used with a 10 minute window.
```

Comment:
There are no guarantees this function will run every 10 minutes, and it's possible it runs every few seconds. Not sure how the cached quota is computed, but can this result in a higher load than expected?

Related: is there a maximum number of monitors for a project/org? Could the rate limits be too high in scenarios with many monitors and result in potential system instability?

Reply:
Sorry, this comment was incorrect, I was toying with a different rate-limit window.

This function only really needs to be run periodically. When a monitor is added for a project we will call `schedule_invalidate_project_config` to update the project config in relay with the newly computed value.